Mitigate Event Loop Errors in Gemini by Adding Connection: close Header

keenranger opened the pull request:
## Title
Mitigate Event Loop Errors in Gemini by Adding Connection: close Header
<!-- e.g. "Implement user authentication feature" -->

## Relevant issues

<!-- e.g. "Fixes #000" -->
#4032 #7667

## Type

<!-- Select the type of Pull Request -->
<!-- Keep only the necessary ones -->

🐛 Bug Fix

## Changes

I am a user of litellm via the zerox library, and I’ve encountered issues that have been widely reported in #7667, and #4032. It appears that sending multiple concurrent requests leads to event loop-related errors.

Upon investigating these issues, it seems that the caching logic for reusing the httpxclient is causing problems under concurrent scenarios. Although I’m not in a position to solve the entire problem, this PR aims to prevent the issue—at least in the Gemini environment I’m using—by applying the workaround suggested by @quangngoc in #4032.

Specifically, as mentioned in [this comment](https://github.com/encode/httpx/issues/2959):

> Adding extra_headers={"Connection": "close"} resolved the issue for me.

In my case, this method has successfully resolved the problem. While it might not be the most elegant solution, it effectively mitigates the failure mode (where the process would otherwise abort completion due to exceptions). I hope this PR is received positively.
<!-- List of changes -->

## [REQUIRED] Testing - Attach a screenshot of any new tests passing locally
If UI changes, send a screenshot/GIF of working UI fixes

<!-- Test procedure -->

keenranger committed:
Add {"Connection": "close"} in gemini header

This commit addresses an issue in the Gemini environment where the caching logic for the httpxclient leads to event loop errors when handling concurrent requests. Based on the workaround suggested by @quangngoc in issue #4032 (see https://github.com/encode/httpx/issues/2959), this change adds the header `{"Connection": "close"}` to mitigate these errors. Although it is not a perfect solution, it prevents the exceptions that were previously causing the completion process to fail.

keenranger commented:
@krrishdholakia I believe you’re already aware of this issue; would you mind reviewing this PR?

krrishdholakia commented:
Thanks @keenranger this is great


Could you add a test for this here - https://github.com/BerriAI/litellm/blob/main/tests/llm_translation/test_vertex.py

keenranger commented:
Hi, @krrishdholakia.
Since this is a change to the existing `completion`, I don’t think it’s necessary to add a new test. Attached is a screenshot of the results from the mentioned test file:
<img width="824" alt="image" src="https://github.com/user-attachments/assets/05aa039a-6f42-4571-9101-addc166a5bf2" />

krrishdholakia commented on litellm/llms/vertex_ai/gemini/vertex_and_google_ai_studio_gemini.py:
```diff
@@ -1053,6 +1053,7 @@ async def async_completion(
             messages=messages,
             optional_params=optional_params,
         )
+        headers["Connection"] = "close"
```
this change should be inside the validate_environment function, instead of in the handler

keenranger replied:
Thank you, it moved to inside the `validate_environment` function
<img width="2304" alt="image" src="https://github.com/user-attachments/assets/de09ee1d-e245-4761-8fae-25b612d1477f" />

keenranger committed:
Move the declaration of the “Connection: close” header into the validate_environment function.

Signed-off-by: Hankyeol Kyung <kghnkl0103@gmail.com>

keenranger commented:
Hi @krrishdholakia, just checking in—if everything looks good, could we go ahead and merge this PR? Thanks!

devdev999 commented:
Hi @krrishdholakia possible to merge this in?

keenranger commented:
Hi @ishaan-jaff 

Since previous attempts to get a review haven’t worked out, could you please take a look and merge it if everything looks good? Your help would be greatly appreciated.

krrishdholakia commented:
Hey @keenranger i don't think this is the right approach for this. 

This defeats the purpose of the cache, by creating a new client on each request. 

We can't accept this PR for now - as it would hurt users in prod, but I do want to resolve this issue. 

Can we plan a better approach here? I'm sorry for the delay in review on my end.

keenranger commented:
@krrishdholakia 
I think the problems are
- THe client is not explicitly specified: from litellm import acompletion.
- This results in retrieving the global client, which seems like a flawed design.

I understand that this is already in production, so making changes recklessly isn’t an option.

What I’d like to ask is:
- Do you plan to maintain this design going forward? Since the current approach calls completion/acompletion directly upon import without a client, it seems unavoidable. However, it also appears problematic from a general Python design perspective.
- Separately, what do you think about adding a parameter to completion/acompletion to disable caching as a temporary solution?

krrishdholakia commented:
Hey @keenranger i believe the issue has a fix discussed - i haven't had a chance to try it yet - https://github.com/BerriAI/litellm/issues/7667#issuecomment-2652652250

Do you think you could see if this solves your problem?

keenranger committed:
Refactor async HTTP client caching to prevent event loop errors

- Introduce AsyncClientCache for per-event-loop client management
- Use weakref to allow event loops to be garbage collected
- Improve client retrieval logic to handle different event loop scenarios
- Remove previous in-memory caching approach
- Add more robust client creation and caching mechanism

Signed-off-by: Hankyeol Kyung <kghnkl0103@gmail.com>

keenranger commented:
Hi @krrishdholakia, I reflected your snippet in the latest commit - Is this the way you wanted? And it seems there's no more error without setting `connection: close` header for my case [zerox](https://github.com/getomni-ai/zerox). But I don't know the detailed logic of your code, would you mind checking if existing logic is not affected by this PR?
- [x] Event loop closed issue resolved
- [ ] Need to check existing cache logic

krrishdholakia commented:
Thank you for the quick fix. Yes - i can create a dev branch to run this through our testing pipeline @keenranger
